src/prometheusstats/prometheus_sink.go

```diff
@@ -0,0 +1,142 @@
+}
+
+type prometheusSinkOption func(sink *prometheusSink)
+
+func WithAddr(addr string) prometheusSinkOption {
+	return func(sink *prometheusSink) {
+		sink.config.addr = addr
+	}
+}
+
+func WithPath(path string) prometheusSinkOption {
+	return func(sink *prometheusSink) {
+		sink.config.path = path
+	}
+}
+
+func WithMapperYamlPath(mapperYamlPath string) prometheusSinkOption {
+	return func(sink *prometheusSink) {
+		sink.config.mapperYamlPath = mapperYamlPath
+	}
+}
+
+// NewPrometheusSink returns a Sink that flushes stats to os.StdErr.
+func NewPrometheusSink(opts ...prometheusSinkOption) gostats.Sink {
+	sink := &prometheusSink{
+		counters:   make(map[string]*prometheus.CounterVec),
+		gauges:     make(map[string]*prometheus.GaugeVec),
+		histograms: make(map[string]*prometheus.HistogramVec),
+		mapper: &mapper.MetricMapper{
+			Registerer: prometheus.DefaultRegisterer,
+		},
+	}
+	for _, opt := range opts {
+		opt(sink)
+	}
+	if sink.config.addr == "" {
+		sink.config.addr = ":9090"
+	}
+	if sink.config.path == "" {
+		sink.config.path = "/metrics"
+	}
+	http.Handle(sink.config.path, promhttp.Handler())
+	go func() {
+		logrus.Infof("Starting prometheus sink on %s%s", sink.config.addr, sink.config.path)
+		_ = http.ListenAndServe(sink.config.addr, nil)
+	}()
+	if sink.config.mapperYamlPath != "" {
+		_ = sink.mapper.InitFromFile(sink.config.mapperYamlPath)
+	} else {
+		_ = sink.mapper.InitFromYAMLString(defaultMapper)
+	}
+	return sink
+}
+
+func (s *prometheusSink) FlushCounter(name string, value uint64) {
+	m, labels, present := s.mapper.GetMapping(name, mapper.MetricTypeCounter)
+	if present {
+		labelNames := make([]string, 0, len(labels))
+		labelValues := make([]string, 0, len(labels))
+		for k, v := range labels {
+			labelNames = append(labelNames, k)
+			labelValues = append(labelValues, v)
+		}
+
+		metricName := mapper.EscapeMetricName(m.Name)
+		if _, ok := s.counters[metricName]; !ok {
+			s.counters[metricName] = promauto.NewCounterVec(prometheus.CounterOpts{Name: metricName}, labelNames)
+		}
+		counter := s.counters[metricName]
+		counter.WithLabelValues(labelValues...).Add(float64(value))
+	}
+}
+
+func (s *prometheusSink) FlushGauge(name string, value uint64) {
+	m, labels, present := s.mapper.GetMapping(name, mapper.MetricTypeCounter)
+	if present {
+		labelNames := make([]string, 0, len(labels))
+		labelValues := make([]string, 0, len(labels))
+		for k, v := range labels {
+			labelNames = append(labelNames, k)
+			labelValues = append(labelValues, v)
+		}
+
+		metricName := mapper.EscapeMetricName(m.Name)
+		if _, ok := s.gauges[metricName]; !ok {
+			s.gauges[metricName] = promauto.NewGaugeVec(prometheus.GaugeOpts{Name: metricName}, labelNames)
+		}
+
+		gauge := s.gauges[metricName]
+		gauge.WithLabelValues(labelValues...).Set(float64(value))
+	}
+}
+
+func (s *prometheusSink) FlushTimer(name string, value float64) {
+	m, labels, present := s.mapper.GetMapping(name, mapper.MetricTypeCounter)
```

Comment:
Why is the type here not `MetricTypeTimer`？

Reply:
cpoy nit